### ISSUE TITLE
(Full Fat) Stateful Service reports System.IO.FileSystem.AccessControl is missing following upgrade to using PackageReferences

### ISSUE THREAD

**Issue by goodisontoffee** [Azure/service-fabric-issues#864]
Struggling with a strange issue which is causing us some headaches.

Here is a really simple repo to demonstrate the issue: https://github.com/goodisontoffee/service-fabric-publishing-issues

If you clone this repo, open VS2017, and Start with Debugging then you'll find that the service comes up correctly and reports and healthy. Without changing anything if you instead decide to Publish the application to your local cluster then you'll start to see the same error message as reported in [639](https://github.com/Azure/service-fabric-issues/issues/639). I have even attempted to add a package reference to the System.IO.FileSystem.AccessControl package referenced in issue 639 (I'm aware one it .net core, and this is full fat framework).

The issue seems to be connected with the different assembly pruning techniques used between "Start with Debugging" and "Package".

If you package the Application then you'll find this in your package output folder and upon publication, this is the set of files is present within the C:\SfDevCluster hierarchy.

Microsoft.ServiceFabric.Data.Interfaces.dll
Microsoft.ServiceFabric.Data.dll
Microsoft.ServiceFabric.Diagnostics.dll
Microsoft.ServiceFabric.Internal.Strings.dll
Microsoft.ServiceFabric.Internal.dll
Microsoft.ServiceFabric.Preview.dll
Microsoft.ServiceFabric.Services.dll
SimpleStatefulService.exe
SimpleStatefulService.exe.config
SimpleStatefulService.pdb
System.Fabric.Management.ServiceModel.dll
System.Fabric.Strings.dll
System.Fabric.dll
System.IO.FileSystem.AccessControl.dll
System.Security.AccessControl.dll
System.Security.Principal.Windows.dll

However, if you "Start with Debugging" then far more is present, new files are in *italics*.

Microsoft.ServiceFabric.Data.Interfaces.dll
Microsoft.ServiceFabric.Data.dll
Microsoft.ServiceFabric.Diagnostics.dll
Microsoft.ServiceFabric.Internal.Strings.dll
Microsoft.ServiceFabric.Internal.dll
Microsoft.ServiceFabric.Preview.dll
Microsoft.ServiceFabric.Services.dll
SimpleStatefulService.exe
SimpleStatefulService.exe.config
SimpleStatefulService.pdb
*System.Data.Common.dll*
*System.Diagnostics.StackTrace.dll*
*System.Diagnostics.Tracing.dll*
*System.Fabric.Management.ServiceModel.XmlSerializers.dll*
System.Fabric.Management.ServiceModel.dll
System.Fabric.Strings.dll
System.Fabric.dll
*System.Globalization.Extensions.dll*
*System.IO.Compression.dll*
System.IO.FileSystem.AccessControl.dll
*System.Net.Http.dll*
*System.Net.Sockets.dll*
*System.Runtime.Serialization.Primitives.dll*
System.Security.AccessControl.dll
*System.Security.Cryptography.Algorithms.dll*
System.Security.Principal.Windows.dll
*System.Security.SecureString.dll*
*System.Threading.Overlapped.dll*
*System.Xml.XPath.XDocument.dll*
*_VS_ServiceFabric_debug.txt*

Some of these assemblies are in place to support the debugging activity but perhaps something that is being excluded is actually required.

If you then look at the original references in the .csproj (prior to upgrading to PackageReferences) you will see that multiple assemblies are referenced from their ..\packages\.... location, but I am uncertain how to replicate this with PackageReferences.

I've also created the same solution using a .net core stateful service, it uses PackageReferences from the start and doesn't exhibit the same issues as the full fat solution moving over to PackageReferences. Sadly, we are unable to adopt .net core for our needs as some of our dependencies do not have a net standard library available for us to make use of. We are planning to move to .net core when we can.

**Comment by ravipal** [comment]
If \<PackageReference\> is used in full framework project and the package contains two different assemblies one from ref and another from lib, then the package will incorrectly include the assembly from lib folder. As a result the application will fail to load the assembly. Please use \<PackageReference\> only for core projects or sdk style project.

If \<PackageReference\> must be used on full framework project, then add the property \<PackageOutputFolder\>true\</PackageOutputFolder\> that will include the righ dependency files in the package.

**Comment by goodisontoffee** [comment]
Okay, thank you for the insight @ravipal. I am still not clear as to why it works when debugging but not when package/publish. If the guidance is that PackageReference is not meant to be used for full framework projects would it be possible to coordinate with this documentation? [Package references (PackageReference) in project files](https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files)

**Comment by goodisontoffee** [comment]
I've added the following section to my .sfproj project file beneath the property group containing the ApplicationProjectTargetsPath element. This appears to have resolved the issue, and I've updated my repo to also contain the fix. It took me a while to discover where to configure the PackageOutputFolder element in the project file but thanks to @ravipal for the pointer. And yes, we are using ApplicationDebugMode="Refresh".

<PropertyGroup>
    <PackageOutputFolder>true</PackageOutputFolder>
</PropertyGroup>
